heketi-cli: Add topology command with options info and load.

humblec opened the pull request:
Example output for topology info command:
[hchiramm@dhcp35-111 go]$ ./heketi-cli topology info

Cluster Id: 0b7c016bf02bab4a9d512b7ffb02923e

Node Id: 08e62b1586526b4a79a349d718010929
Cluster Id: 0b7c016bf02bab4a9d512b7ffb02923e
Zone: 1
Management Hostname: dhcp35-111.lab.eng.blr.redhat.com
Storage Hostname: dhcp35-111.lab.eng.blr.redhat.com
Devices:
    Id:b320790fc222215091a82f4747e5bba9   Name:/dev/sde1           Size (GiB):500     Used (GiB):84      Free (GiB):415
        Bricks:
            Id:163aca2d2eb00726e6655cba174e9d09   Size (GiB):16      Path: /mockpath
            Id:f753ef2b94e09c2f5e8cae2e8a63a889   Size (GiB):16      Path: /mockpath

Node Id: 75d1d22afb654ba90513706c2514aa00
Cluster Id: 0b7c016bf02bab4a9d512b7ffb02923e
Zone: 1
Management Hostname: dhcp35-112.lab.eng.blr.redhat.com
Storage Hostname: dhcp35-112.lab.eng.blr.redhat.com
Devices:
    Id:74e9e2a3d9732d41d60f2f63082665c7   Name:/dev/sde1           Size (GiB):500     Used (GiB):95      Free (GiB):404
        Bricks:
            Id:08c972041a786d48473e45e186252008   Size (GiB):5       Path: /mockpath
            Id:2151ee4d4685b7463543e04a67ca445d   Size (GiB):16      Path: /mockpath

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

lpabon commented:
Hi @humblec , the data just needs to be formatted a bit.  For example:

```
Cluster Id: 0b7c016bf02bab4a9d512b7ffb02923e
    \tNode Id: 08e62b1586526b4a79a349d718010929
        \t\tZone:

humblec commented:
@lpabon its already intended as shown below

[hchiramm@dhcp35-111 go]$ ./heketi-cli topology info
Cluster Id: 0b7c016bf02bab4a9d512b7ffb02923e

	Node Id: 08e62b1586526b4a79a349d718010929
	Cluster Id: 0b7c016bf02bab4a9d512b7ffb02923e
	Zone: 1
	Management Hostname: dhcp35-111.lab.eng.blr.redhat.com
	Storage Hostname: dhcp35-111.lab.eng.blr.redhat.com
	Devices:
		Id:b320790fc222215091a82f4747e5bba9   Name:/dev/sde1           Size (GiB):500     Used (GiB):84      Free (GiB):415     
			Bricks:
				Id:163aca2d2eb00726e6655cba174e9d09   Size (GiB):16      Path: /mockpath
				Id:192b4727f144173885b7d85e0a7e6157   Size (GiB):5       Path: /mockpath
				Id:59c8f3e37eebb9b4a985bbcd7de60848   Size (GiB):7       Path: /mockpath
				Id:78fa94a14ec2acaafbe5342e3f2dd039   Size (GiB):7       Path: /mockpath
				Id:9e0d656efc677eb7f31af8e493f50e7e   Size (GiB):16      Path: /mockpath
				Id:f4a61f2e914bff7ca09c6bdb8f144969   Size (GiB):16      Path: /mockpath
				Id:f753ef2b94e09c2f5e8cae2e8a63a889   Size (GiB):16      Path: /mockpath

	Node Id: 75d1d22afb654ba90513706c2514aa00
	Cluster Id: 0b7c016bf02bab4a9d512b7ffb02923e
	Zone: 1
	Management Hostname: dhcp35-112.lab.eng.blr.redhat.com
	Storage Hostname: dhcp35-112.lab.eng.blr.redhat.com
	Devices:
		Id:74e9e2a3d9732d41d60f2f63082665c7   Name:/dev/sde1           Size (GiB):500     Used (GiB):95      Free (GiB):404     
			Bricks:
				Id:08c972041a786d48473e45e186252008   Size (GiB):5       Path: /mockpath
				Id:2151ee4d4685b7463543e04a67ca445d   Size (GiB):16      Path: /mockpath
				Id:38c44747be82eed060b114002adf0a5d   Size (GiB):5       Path: /mockpath
				Id:5d546d453f7dd9633d45fa7a9e09974c   Size (GiB):5       Path: /mockpath
				Id:73cae4c17dd107021bc0202941ea1046   Size (GiB):16      Path: /mockpath
				Id:8be9c1c4c4347c7d10dd000981978de4   Size (GiB):16      Path: /mockpath
				Id:94fedb4d0eb86a84922b554d558680ed   Size (GiB):16      Path: /mockpath
				Id:ba9de783b61d73e71fda396b10f1b2c2   Size (GiB):16      Path: /mockpath

lpabon commented:
@humblec  Ah sweet.  Yeah, that was just github comment formatting issue.  Ok I will try it out and review the change. tonight.

lpabon commented:
Human format looks good.  I have a few questions on the JSON format:

```json
{
  "clusters": [
    {
      "id": "b8144fdef38806c9db843afe1d14316f",
      "nodes": {
        "Nodes": [
          {
            "nodeinfo": {
              "zone": 2,
              "hostnames": {
                "manage": [
                  "192.168.10.101"
                ],
                "storage": [
                  "192.168.10.101"
                ]
              },
              "cluster": "b8144fdef38806c9db843afe1d14316f"
```
* Is there a reason why you have `cluster` inside `nodes`?
* I don't think it is necessary to have ` "nodes" : { "Nodes : [  "nodeinfo" `, you could probably have `"nodes" : [  { ....node information here.. }, { ... second node here ... }...]`  Like `"nodes": [ { "zone" : 2, "hostnames" : {...}... etc}, {...}]`.  

What do you think?

lpabon commented on apps/glusterfs/models.go:
```diff
@@ -25,6 +25,25 @@ import (
+type Topo struct {
+	ClusterList []Cluster `json:"clusters"`
+}
+
```
These should only be in the server models.go directory if they are used for REST commands.  Instead, create a models file for the heketi-cli

humblec replied:
@lpabon 'cluster' or 'id of cluster' is part of nodeinfo response. Thats why its available in nodeinfo field. Considering the fact that its just one ID field, I feel its fine, so that we return the node information intact for some one to make use of it, if needed. Thats my thought. Please revert if you have any comments. Regarding 'models.go' for cli, I thought about it before placing it in server models.  but do u think at any stage we may have to create a route for topology in server side? in that case, this may be a better place. What do you think?

lpabon commented on client/cli/go/cmds/topology.go:
```diff
@@ -109,3 +117,75 @@ var loadCommand = &cobra.Command{
+
+			for i, _ := range topoinfo.ClusterList {
+
+				fmt.Fprintf(stdout, "Cluster Id: %v\n", topoinfo.ClusterList[i].Id)
```
Very cool.  Just break add a few more comments to make easier to read.

humblec replied:
@lpabon Thanks :) . Sure I will add more comments.

lpabon commented:
@MohamedAshiqrh Please review when you can the cobra changes

MohamedAshiqrh commented:
Verified cobra changes, looks good to me. good work :+1: .

humblec commented:
@lpabon can you please confirm whether we need seperate client model.go ? or can you please respond to last comment from me ?

lpabon commented:
Ok, a couple of things.  I tend to agree with your comment about keeping it in `apps/glusterfs/models.go` so thinking along those lines, we should had a new issue (for Release 3 - future) which adds a Topology endpoint to the API, where one can add or retrieve topology information.

For now we should add the structs as you mentioned to `models.go`, but here is a simplified version:

```go
// models.go
// Topology

type TopologyCluster struct {
	Volumes []VolumeInfoResponse `json:"volumes"`
	Nodes   []NodeInfoResponse   `json:"nodes"`
	Id      string               `json:"id"`
}

type TopologyInfoResponse struct {
	Clusters []TopologyCluster `json:"clusters"`
}
```

This will output in JSON the entire topology without unneeded members.

I think it is ok to have `heketi-cli` build this now, and Release-3 we will replace it with the API.  The CLI will still work the same.

humblec commented:
@lpabon thanks for the update. Are you saying, the newly introduced  structures in the patch models.go  has to be replaced with the structures you mentioned in your last commet? or its for  next release-3 version ? I am bit confused with this , "I think it is ok to have heketi-cli build this now, and Release-3 we will replace it with the API. The CLI will still work the same." so the question.

lpabon commented:
@humblec Sorry for the confusion.  The proposed structure above should be used in this patch.  The only thing that will be done for Release 3 is a new API endpoint.

humblec commented:
@lpabon nw, I will adopt above mentioned structure in new patch, however the only difference I see here is extra 'volumelist' in the topology output. The 'cluster id' field will still be part of nodeinfo. It should be fine any way.  thanks!

humblec commented:
@lpabon please review new PR with Volume Info response.

lpabon commented on client/api/go-client/topology.go:
```diff
@@ -0,0 +1,71 @@
+//
```
This file needs unit tests

lpabon commented on client/api/go-client/topology.go:
```diff
@@ -0,0 +1,71 @@
+import "github.com/heketi/heketi/apps/glusterfs"
+
+func (c *Client) TopologyInfo() (*glusterfs.TopologyInfoResponse, error) {
+	var topo glusterfs.TopologyInfoResponse
```
I'm not sure how this worked, but the list topo.ClusterList needs to be initialized.  Here is one way

```go
topo := glusterfs.TopologyInfoResponse {
    ClusterLIst: make([]glusterfs.Cluster, 0)
}
```

lpabon commented on client/api/go-client/topology.go:
```diff
@@ -0,0 +1,71 @@
+		}
+
+		var cluster glusterfs.Cluster
+		cluster.Id = clusteri.Id
```
This works, here is another way:

```go
cluster := glusterfs.Cluster {
    Id: clusteri.Id
}
```

lpabon commented on client/api/go-client/topology.go:
```diff
@@ -0,0 +1,71 @@
+		var cluster glusterfs.Cluster
+		cluster.Id = clusteri.Id
+
+		//	var volume glusterfs.VolumeInfoResponse
```
Please remove if it is not needed.

lpabon commented on client/api/go-client/topology.go:
```diff
@@ -0,0 +1,71 @@
+			return nil, err
+		}
+
+		for _, volumes := range volumelist.Volumes {
```
Add comments

lpabon commented on client/api/go-client/topology.go:
```diff
@@ -0,0 +1,71 @@
+
+				cluster.Volumes = append(cluster.Volumes, *volumesi)
+
+			}
```
Remove empty lines

lpabon commented on client/api/go-client/topology.go:
```diff
@@ -0,0 +1,71 @@
+		cluster.Id = clusteri.Id
+
+		//	var volume glusterfs.VolumeInfoResponse
+		volumelist, err := c.VolumeList()
```
This one is not necessary which caused you to add Line 50.  Instead, `clusteri` has the list of volumes already, and you can use that list in line 43.

lpabon commented on client/api/go-client/topology.go:
```diff
@@ -0,0 +1,71 @@
+			}
+
+			cluster.Nodes = append(cluster.Nodes, *nodei)
+
```
Please remove empty lines

lpabon commented on client/api/go-client/topology.go:
```diff
@@ -0,0 +1,71 @@
+			}
+		}
+
+		for _, node := range clusteri.Nodes {
```
add comments

lpabon commented on client/api/go-client/topology.go:
```diff
@@ -0,0 +1,71 @@
+
+		topo.ClusterList = append(topo.ClusterList, cluster)
+	}
+	return &topo, nil
```
change to `topo`

lpabon commented on client/cli/go/cmds/topology.go:
```diff
@@ -0,0 +1,260 @@
+		heketi := client.NewClient(options.Url, options.User, options.Key)
+
+		// Create Topology
+
```
Please remove empty line

lpabon commented on client/cli/go/cmds/topology.go:
```diff
@@ -0,0 +1,260 @@
+				fmt.Fprintf(stdout, "\nCluster Id: %v\n", topoinfo.ClusterList[i].Id)
+				fmt.Fprintf(stdout, "\n    %s\n", "Volumes:")
+				for k, _ := range topoinfo.ClusterList[i].Volumes {
+					// Format and print volumeinfo  on this cluster
```
please add an empty line above the comment

humblec commented:
@lpabon if I didnt miss anything,  I have addressed all the comments. The only one left here is the unit test. As @MohamedAshiqrh is working on tests for all the cli commands, he is adding the unit test for topology command as well in his patch.

lpabon commented:
@humblec @MohamedAshiqrh please add the unit test for this in this patch, not in another.  That way we continue to test as we develop the code

lpabon commented:
@humblec Very nice changes, all we need is the unit test.  The unit test is not for the cli, that is what I think @MohamedAshiqrh is working on.  The unit test I am asking for is for the client api call Topology() which you created.  There is a file already which has all the client unit tests, you can add your unit test function to [client_test.go](https://github.com/heketi/heketi/blob/master/client/api/go-client/client_test.go)

MohamedAshiqrh commented:
@lpabon @humblec I will take care of cli-test part of topology command. Just making sure we are all in same page.:)

lpabon commented on client/api/go-client/client_test.go:
```diff
@@ -52,6 +53,59 @@ func setupHeketiServer(app *glusterfs.App) *httptest.Server {
+	topology, err = c.TopologyInfo()
+	tests.Assert(t, err == nil)
+	tests.Assert(t, topology.ClusterList[0].Nodes[0].Id == node.Id)
+
```
You probably want to add more nodes and devices and volumes to correctly test the topology info.  You also would want to test it with multiple clusters.

humblec replied:
@lpabon considering we need this patch soon can you please merge it ? I will add more test scenarios later, hope its fine.

lpabon replied:
@humblec The patch is not complete without testing, therefore it cannot be merged.

lpabon commented on client/api/go-client/client_test.go:
```diff
@@ -52,6 +53,59 @@ func setupHeketiServer(app *glusterfs.App) *httptest.Server {
+	tests.Assert(t, reflect.DeepEqual(nodeReq.Hostnames, node.Hostnames))
+	tests.Assert(t, len(node.DevicesInfo) == 0)
+
+	//// Get information about the Topology and verify the node add
```
Just a simple thing, the line has four `/` instead of two

lpabon commented:
It looks like it is working according to the CI system.  Great job fixing it.

humblec commented:
Thanks @lpabon !

lpabon commented:
Closes #276
